q2_assembly/helpers/helpers.py

```diff
@@ -0,0 +1,91 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import os
+import warnings
+
+import numpy as np
+from q2_types.bowtie2 import Bowtie2IndexDirFmt
+from q2_types_genomics.per_sample_data import BAMDirFmt, ContigSequencesDirFmt
+from qiime2.util import duplicate
+
+
+def partition_contigs(
+    contigs: ContigSequencesDirFmt, num_partitions: int = None
+) -> ContigSequencesDirFmt:
+    partitioned_contigs = {}
+
+    contigs = list(contigs.path.iterdir())
+    num_samples = len(contigs)
+    if num_partitions is None:
+        num_partitions = num_samples
+    elif num_partitions > num_samples:
+        warnings.warn(
+            "You have requested a number of partitions"
+            f" '{num_partitions}' that is greater than your number"
+            f" of samples '{num_samples}.' Your data will be"
+            f" partitioned by sample into '{num_samples}'"
+            " partitions."
+        )
+        num_partitions = num_samples
+
+    # If the number of splits to make is larger than the collection given,
+    # np.array_split partitions the collection into individual elements
+    contigs = np.array_split(contigs, num_partitions)
+    for i, samples in enumerate(contigs, 1):
+        result = ContigSequencesDirFmt()
+
+        for sample_fp in samples:
+            # These paths are defined in the ContigSequencesDirFmt class as
+            # {sample_id}_contigs.(fa | fasta). This should get the id from a
+            # name like that
+            sample_id = sample_fp.name.rsplit("_contigs", 1)[0]
+            duplicate(sample_fp, result.path / sample_fp.name)
+
+        if num_partitions == num_samples:
+            partitioned_contigs[sample_id] = result
+        else:
+            partitioned_contigs[i] = result
+
+    return partitioned_contigs
```

Comment:
QIIME knows this action is supposed to return a collection because in plugin_setup the semantic type of the return is `Collection[SampleData[Contigs]]`. The view type of the return is just the view type of one element of the collection so the view type annotation is just `ContigSequencesDirFmt`.

If we were doing things 100% accurately, the view type annotation should be something like `Dict[String, ContigSequencesDirFmt]`, but we don't currently support view type annotations that are nested like that. ALL output collections are dicts and they are always keyed on strings, so there is no ambiguity in the view type annotation when you know that, but it does require you to know that extra bit of information.

Reply:
Gotcha, thanks for the explanation! 🙇